Crash when entering the `º` char while on multicursor

https://github.com/user-attachments/assets/1f1953da-b2e1-4396-9e2c-df1157702618

Need a bit of help to give you more info on this. I found it while batch renaming a variable with ⌘D in a TypeScript file, where I accidentally pressed `º` and Zed panicked instantly. I could reproduce it on several of my existing projects, but not on a new NodeJS project, so it must have something to do with specific TypeScript / ESLint versions, maybe? Let me know any other logs or info you might need.

<details>
<summary>zed.log</summary>
<pre>
2025-02-19T08:38:56.461189+01:00 [ERROR] {
  "thread": "main",
  "payload": "byte index 5 is not a char boundary; it is inside 'º' (bytes 4..6) of `myElº (parameter) myElº: { foo: string; }`",
  "location_data": {
    "file": "crates/language/src/language.rs",
    "line": 1783
  },
  "backtrace": [
    "backtrace::capture::Backtrace::create::h71b44b128423336c+10137520",
    "backtrace::capture::Backtrace::new::h743343b0418d06db+10137332",
    "zed::reliability::init_panic_hook::{{closure}}::hebfbc38617d4941e+84560180",
    "std::panicking::rust_panic_with_hook::h09e8a656f11e82b2+65514672",
    "std::panicking::begin_panic_handler::{{closure}}::h1230eb3cc91b241c+65511768",
    "std::sys::backtrace::__rust_end_short_backtrace::hc3491307aceda2c2+65503872",
    "_rust_begin_unwind+65510956",
    "core::panicking::panic_fmt::ha4b80a05b9fff47a+94193848",
    "core::str::slice_error_fail_rt::h7cb1d51121ce60d3+16876860",
    "core::str::slice_error_fail::he72416048b197ca9+94195252",
    "project::lsp_store::LspStore::resolve_completions::{{closure}}::{{closure}}::h1aac5b4a19e621fc+51743424",
    "async_task::raw::RawTask<F,T,S,M>::run::hbd4551c4797c6092+47788928",
    "gpui::platform::mac::dispatcher::trampoline::ha2d92e1182005862+33280928",
    "<gpui::platform::mac::platform::MacPlatform as gpui::platform::Platform>::run::h181325e9ddacf5d5+33343172",
    "gpui::app::Application::run::ha034bbb7202bfcac+82337600",
    "zed::main::hd6d57659e5c51328+85079864",
    "std::sys::backtrace::__rust_begin_short_backtrace::h9226acd42c3dab52+83853932",
    "std::rt::lang_start::{{closure}}::h5e0f06121f73f53d+84286184",
    "std::rt::lang_start_internal::hdd117cb81a316264+65445352",
    "_main+85090104"
  ],
  "app_version": "0.173.11",
  "app_commit_sha": "44350ebca517498f7695d51f285e16a163fdafcc",
  "release_channel": "stable",
  "target": "aarch64-apple-darwin",
  "os_name": "macOS",
  "os_version": "15.1.1",
  "architecture": "aarch64",
  "panicked_on": 1739950736461,
  "system_id": "46ff795d-a2f1-4936-b8c7-c43d3ffcfbef",
  "installation_id": "4e8b54a0-9c9f-46ee-a035-6665003b99e2",
  "session_id": "b90341be-c813-4a4d-8702-28ea6af25f4f"
}
</pre>
</details>

Zed: v0.173.11
OS: macOS 15.1.1 (24B91)
Memory: 8GB
Architecture: ARM64

--- a/crates/project/src/lsp_store.rs
+++ b/crates/project/src/lsp_store.rs
@@ -8944,7 +8944,7 @@ fn ensure_uniform_list_compatible_label(label: &mut CodeLabel) {
             }
             _ => {
                 new_text.push(c);
-                new_idx += 1;
+                new_idx += c.len_utf8();
                 last_char_was_space = false;
             }
         }
@@ -9015,27 +9015,16 @@ fn ensure_uniform_list_compatible_label(label: &mut CodeLabel) {
 }
 
 #[cfg(test)]
-#[test]
-fn test_glob_literal_prefix() {
-    assert_eq!(glob_literal_prefix(Path::new("**/*.js")), Path::new(""));
-    assert_eq!(
-        glob_literal_prefix(Path::new("node_modules/**/*.js")),
-        Path::new("node_modules")
-    );
-    assert_eq!(
-        glob_literal_prefix(Path::new("foo/{bar,baz}.js")),
-        Path::new("foo")
-    );
-    assert_eq!(
-        glob_literal_prefix(Path::new("foo/bar/baz.js")),
-        Path::new("foo/bar/baz.js")
-    );
+mod tests {
+    use language::HighlightId;
 
-    #[cfg(target_os = "windows")]
-    {
-        assert_eq!(glob_literal_prefix(Path::new("**\\*.js")), Path::new(""));
+    use super::*;
+
+    #[test]
+    fn test_glob_literal_prefix() {
+        assert_eq!(glob_literal_prefix(Path::new("**/*.js")), Path::new(""));
         assert_eq!(
-            glob_literal_prefix(Path::new("node_modules\\**/*.js")),
+            glob_literal_prefix(Path::new("node_modules/**/*.js")),
             Path::new("node_modules")
         );
         assert_eq!(
@@ -9043,8 +9032,43 @@ fn test_glob_literal_prefix() {
             Path::new("foo")
         );
         assert_eq!(
-            glob_literal_prefix(Path::new("foo\\bar\\baz.js")),
+            glob_literal_prefix(Path::new("foo/bar/baz.js")),
             Path::new("foo/bar/baz.js")
+        );
+
+        #[cfg(target_os = "windows")]
+        {
+            assert_eq!(glob_literal_prefix(Path::new("**\\*.js")), Path::new(""));
+            assert_eq!(
+                glob_literal_prefix(Path::new("node_modules\\**/*.js")),
+                Path::new("node_modules")
+            );
+            assert_eq!(
+                glob_literal_prefix(Path::new("foo/{bar,baz}.js")),
+                Path::new("foo")
+            );
+            assert_eq!(
+                glob_literal_prefix(Path::new("foo\\bar\\baz.js")),
+                Path::new("foo/bar/baz.js")
+            );
+        }
+    }
+
+    #[test]
+    fn test_multi_len_chars_normalization() {
+        let mut label = CodeLabel {
+            text: "myElˇ (parameter) myElˇ: {\n    foo: string;\n}".to_string(),
+            runs: vec![(0..6, HighlightId(1))],
+            filter_range: 0..6,
+        };
+        ensure_uniform_list_compatible_label(&mut label);
+        assert_eq!(
+            label,
+            CodeLabel {
+                text: "myElˇ (parameter) myElˇ: { foo: string; }".to_string(),
+                runs: vec![(0..6, HighlightId(1))],
+                filter_range: 0..6,
+            }
         );
     }
 }

